File: ts2/position.py
#
#   Copyright (C) 2008-2013 by Nicolas Piganeau                                
#   npi@m4x.org                                                           
#                                                                         
#   This program is free software; you can redistribute it and/or modify  
#   it under the terms of the GNU General Public License as published by  
#   the Free Software Foundation; either version 2 of the License, or     
#   (at your option) any later version.                                   
#                                                                         
#   This program is distributed in the hope that it will be useful,       
#   but WITHOUT ANY WARRANTY; without even the implied warranty of        
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         
#   GNU General Public License for more details.                          
#                                                                         
#   You should have received a copy of the GNU General Public License     
#   along with this program; if not, write to the                         
#   Free Software Foundation, Inc.,                                       
#   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             
#

from PyQt4.QtGui import *

class Position:
    """Class Position
    TODO Document Position class"""

    def __init__(self, trackItem = None, previousTI = None, positionOnTI = 0.0):
        self._trackItem = trackItem
        self._previousTI = previousTI
        self._positionOnTI = positionOnTI

    @property
    def trackItem(self):
        return self._trackItem

    @property
    def previousTI(self):
        return self._previousTI

    @property
    def positionOnTI(self):
        return float(self._positionOnTI)

    def next(self, pos = 0, direction = -1):
        return Position(self._trackItem.getFollowingItem(self._previousTI, direction),\
                        self._trackItem, \
                        pos)
    
    def previous(self, pos = None):
        if pos is None:
            pos = self._previousTI.realLength
        return Position(self._previousTI,\
                        self._previousTI.getFollowingItem(self._trackItem),\
                        pos)

    def distanceToPosition(self, p):
        """Returns the distance between the current position and p if p is ahead of 
        current position. Returns 0 otherwise."""
        res = 0.0
        if self.trackItem != p.trackItem:
            res = self.trackItem.realLength - self.positionOnTI
            cur = self.next()
            while cur.trackItem != p.trackItem:
                if cur.trackItem is None:
                    return -1
                res += cur.trackItem.realLength
                cur = cur.next()
            res += p.positionOnTI
        else:
            res = p.positionOnTI - self.positionOnTI
        return max(res, 0)

        
    def trackItemsToPosition(self, p):
        """Returns a list of all the trackItems between this position and position p.
        including the trackItem of this position and the trackItem of position p."""
        til = []
        cur = self
        while cur.trackItem != p.trackItem and not cur.isOut():
            til.append(cur.trackItem)
            cur = cur.next()
        til.append(p.trackItem)
        return til

    def isOut(self):
        if self.trackItem.tiType.startswith("E"):
            return True
        else:
            return False

    def reverse(self):
        """Returns a position that is physically on the exact same place, but
        coming from the opposite direction"""
       	self._positionOnTI = self._trackItem.realLength - self._positionOnTI
        self._previousTI = self._trackItem.getFollowingItem(self._previousTI)
        return Position(self._trackItem, self._previousTI, self._positionOnTI)

    def __eq__(self, p):
        return (self._trackItem == p.trackItem \
            and self._previousTI == p.previousTI \
            and self._positionOnTI == p.positionOnTI)

    def __ne__(self, p):
        return not (self == p)

    def __add__(self, length):
        if self._positionOnTI + length < self._trackItem.realLength:
            return Position(self._trackItem, \
                            self._previousTI, \
                            self._positionOnTI + length)
        else:
            return self.next() + (length + self._positionOnTI - self._trackItem.realLength)

    def __sub__(self, length):
        if (self._positionOnTI - length > 0):
            return Position(self._trackItem, \
                            self._previousTI, \
                            self._positionOnTI - length)
        else:
            return self.previous(self._previousTI.realLength) - \
                   (length - self._positionOnTI)

    def __iadd__(self, length):
        self = self + length
        return self

    def __isub__(self, length):
        self = self - length
        return self
    

File: server/simulation/position.go
/*   Copyright (C) 2008-2016 by Nicolas Piganeau and the TS2 TEAM
 *   (See AUTHORS file)
 *
 *   This program is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation; either version 2 of the License, or
 *   (at your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program; if not, write to the
 *   Free Software Foundation, Inc.,
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */

package simulation

import (
	//	"encoding/json"
	"fmt"
)

/*
A Position object is a point on a TrackItem.

A Position is defined as being positionOnTI meters away from the end of this
TrackItem that is connected to PreviousItem.

Note that a Position has a direction, so that for any point on a TrackItem,
there are two Positions that can be defined:

- one starting from one end of the TrackItem.
- the other starting from the other end.

You can get the other Position by calling Reversed()
*/
type Position struct {
	TrackItem    TrackItem
	PreviousItem TrackItem
	PositionOnTI float64
}

/*
IsValid returns true if this is a valid position (i.e. items are connected, and
distance is positive), false otherwise.
*/
func (pos Position) IsValid() bool {
	if !pos.TrackItem.IsConnected(pos.PreviousItem) {
		return true
	}
	return false
}

// IsOut() is true if this position is out of the scene and moving forward
func (pos Position) IsOut() bool {
	if pos.TrackItem.Type() == "EndItem" && pos.PreviousItem != nil {
		return true
	}
	return false
}

// Position.Next() is the Position on the next TrackItem with regard to this Position
func (pos Position) Next(dir Direction) Position {
	nextTi, _ := pos.TrackItem.FollowingItem(pos.PreviousItem, dir)
	return Position{nextTi, pos.TrackItem, 0}
}

/*
Reversed returns the position that is at the same position but in the
opposite direction.
*/
func (pos Position) Reversed() Position {
	ti := pos.TrackItem
	pti := pos.PreviousItem
	nti, _ := ti.FollowingItem(pti, 0)
	distance := pos.TrackItem.RealLength() - pos.PositionOnTI
	return Position{ti, nti, distance}
}

/*
NewPosition returns a pointer to a Position defined by the given Simulation
pointer and PositionRepr.
*/
func NewPosition(sim *Simulation, posRepr PositionRepr) (*Position, error) {
	ti, ok := sim.TrackItems[posRepr.TrackItemId]
	if !ok {
		return nil, fmt.Errorf("Unknown item with ID: %i", posRepr.TrackItemId)
	}
	pti, ok := sim.TrackItems[posRepr.PreviousItemId]
	if !ok {
		return nil, fmt.Errorf("Unknown item with ID: %i", posRepr.PreviousItemId)
	}
	newPos := Position{ti, pti, posRepr.PositionOnTi}
	if !newPos.IsValid() {
		return nil, fmt.Errorf("Position (%i, %i, %f) is not valid", posRepr.TrackItemId, posRepr.PreviousItemId, posRepr.PositionOnTi)
	}
	return &newPos, nil
}

/*
PositionRepr is a representation of a Position that is independant of a Simulation
Object.
*/
type PositionRepr struct {
	TrackItemId    int     `json:"trackItem"`
	PreviousItemId int     `json:"previousTI"`
	PositionOnTi   float64 `json:"positionOnTI"`
}
